src/driver-marsgaming/marsgaming-protocol.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+#include "marsgaming-buttons.h"
+#include "marsgaming-leds.h"
+
+enum marsgaming_report_type {
+	MARSGAMING_MM4_REPORT_TYPE_UNKNOWN_1 = 0x01,
+	MARSGAMING_MM4_REPORT_TYPE_WRITE = 0x02,
+	MARSGAMING_MM4_REPORT_TYPE_READ = 0x03,
+	MARSGAMING_MM4_REPORT_TYPE_UNKNOWN_4 = 0x04,
+	MARSGAMING_MM4_REPORT_TYPE_UNKNOWN_6 = 0x06
+};
+
+struct marsgaming_report_resolution_info {
+	bool enabled;
+	uint16_t x_res;
+	uint16_t y_res;
+	// 4 lowest bits, each one corresponds to one led
+	// Resolution 0 -> b0000
+	// Resolution 1 -> b0001
+	// Resolution 2 -> b0011
+	// etc
+	uint8_t led_bitset;
+	uint8_t zeros_3[2];
+} __attribute__((packed));
+
+struct marsgaming_report_resolutions {
+	uint8_t usb_report_id;
+	uint8_t report_type; // marsgaming_report_type
+	uint8_t unknown_2; // 0x4f
+	uint8_t profile_id;
+	uint8_t unknown_4; // 0x2a
+	uint8_t unknown_5; // 0x00
+	uint8_t unknown_6; // 0x00 from device | 0xfa from host
+	uint8_t unknown_7; // 0x00 from device | 0xfa from host
+	uint8_t count_resolutions;
+	uint8_t current_resolution;
+	union {
+		struct marsgaming_report_resolution_info resolutions[6];
+		uint8_t padding[54];
+	};
+} __attribute__((packed));
+
+_Static_assert(sizeof(struct marsgaming_report_resolutions) == 64, "Marsgaming resolution report is not 64 bytes long");
+
+struct marsgaming_report_buttons {
+	uint8_t usb_report_id;
+	uint8_t report_type; // marsgaming_report_type
+	uint8_t unknown_2; // 0x90
+	uint8_t profile_id;
+	uint8_t unknown_4; // 0x4d
+	uint8_t unknown_5; // 0x00
+	uint8_t unknown_6; // 0x00 from device | 0xfa from host
+	uint8_t unknown_7; // 0x00 from device | 0xfa from host
+	uint8_t button_count;
+	union {
+		struct marsgaming_button_info buttons[253];
+		uint8_t padding[1015];
+	};
```

Comment:
Same as the above, these fields have different sizes, and it also seems like this doesn't have to be a union at all.

Reply:
The padding is so that the object size is exactly 1024. An array of `marsgaming_button_info` does not fit evenly inside the report, there would be 3 bytes left over, but the report still has to be 1024 to avoid reading or writing to unallocated memory. Is there any other way of doing this?

Reply:
You can remove the union and just add padding at the end:
```c
	struct marsgaming_button_info buttons[253];
	uint8_t padding[3];
```

By the way, I think this struct is also missing ` __attribute__((packed))` :)